### PR TITLE
[WIP] Hackdays project: particle system layer

### PR DESCRIPTION
**Pull request by endanke**
This PR is not planned to be merged, but we will share some development notes here.

### Proposal: Particle System Layer

A new layer type which makes it easy to display large amount of repeating content on the map with random variation on several attributes.

Potential use-cases:
- Weather conditions: clouds, snow, rain, forest fires, wind movement
- City night lights on dark styles
- Data visualization (eg.: population density)
- ...

Screenshot 2022-01-17 at 11 30 41

On the example above, the snow particles are defined with the following layer configuration:
```
map.addLayer({
    "id": "particles-snow",
    "type": "particle",
    "source": "snow-budapest",
    "paint": {
        "particle-radius": 0.2,
        "particle-emitter-type": "gradient",
        "particle-color-start": "#ffffff",
        "particle-color-end": "#ffffff",
        "particle-emitter-ttl-min": 1000,
        "particle-emitter-ttl-max": 5000,
        "particle-emitter-velocity-min": 0.1,
        "particle-emitter-velocity-max": 0.3,
        "particle-emitter-direction": [-0.005, 0, -1],
        "particle-emitter-velocity-min": 10,
        "particle-emitter-velocity-max": 20,
        "particle-emitter-elevation-min": 8000,
        "particle-emitter-elevation-max": 9000,
        "particle-emitter-offset-min": 0,
        "particle-emitter-offset-max": 50
    }
});
```
(Note: there are several other hardcoded configuration values in this demo which is not yet exposed in the layer properties, but can be added in a full-featured implementation.) 

Known bugs and issues:
- At high pitch values random particles appear at the horizon
- Particles are incorrectly scaled when changing zoom levels
- The particle system is stored statically and not emptied
- Heavy calculations in the draw calls, should be moved to a worker thread
- Images are loaded from a raster layer
- Particles are not ordered correctly on the z axis
- The layer is not drapeable, when terrain is enabled the city lights are displayed over water and other layers
- ...

Known things to do:
- Fix bugs and clean up the code
- Add more paint properties to support the styling of all emitter parameters
- Add support for image source
- Add more examples to showcase further use cases: rain, snow, fireworks, etc.
- ...

## Launch Checklist

 - [ ] briefly describe the changes in this PR
 - [ ] include before/after visuals or gifs if this PR includes visual changes
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [ ] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [ ] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog></changelog>`

### CODE REVIEW

File: src/render/draw_particle.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            if (!emitter.tileId.equals(bucket.tileId)) {
+                continue;
+            }
+            if (emitter.paint.get('particle-emitter-type') != layer.paint.get('particle-emitter-type')) {
+                continue;
+            }
+            for (const particle of emitter.particles) {
+                particleCount += 1;
+                particlePositions.push([
+                    emitter.location.x + emitter.zoom * particle.locationOffset.x,
```
Comment (endanke) [inline src/render/draw_particle.js:92]: Using `emitter.zoom` here is incorrect and causes issues when changing zoom levels. We would need to revisit how we store the location for the emitters and particles.

**Comment by brncsk** [comment]
Hey, @endanke, an offtopic question: any reason you chose Budapest for the demo? nice to see my hometown for this cool demo ;)

**Comment by joedjc** [comment]
This looks really interesting @endanke...would there be any plans to merge this in future?

**Comment by endanke** [comment]
@joedjc Happy to hear you like it! At the moment it's just a proof of concept, there are no plans to merge it yet. But I hope there'll be enough customer interest in the future to get this into production.

@brncsk I grew up there as well! :)

**Pull request closed without merging**